src/me.rs

```diff
@@ -746,29 +758,53 @@ pub fn estimate_motion_ss2<T: Pixel>(
 
     for omv in pmvs.iter() {
       if let Some(pmv) = omv {
-        let x_lo = po.x + (((pmv.col as isize / 8 - range).max(mvx_min / 8).min(mvx_max / 8)) >> 1);
-        let x_hi = po.x + (((pmv.col as isize / 8 + range).max(mvx_min / 8).min(mvx_max / 8)) >> 1);
-        let y_lo = po.y + (((pmv.row as isize / 8 - range).max(mvy_min / 8).min(mvy_max / 8)) >> 1);
-        let y_hi = po.y + (((pmv.row as isize / 8 + range).max(mvy_min / 8).min(mvy_max / 8)) >> 1);
+        if fi.config.speed_settings.diamond_me {
+          let mut predictors = get_subset_predictors::<T>(
+            &bo_adj_h,
+            MotionVector{row: pmv.row, col: pmv.col},
+            fi.w_in_b, fi.h_in_b,
+            &frame_mvs, frame_ref, 0
+          );
 
-        full_search(
-          x_lo,
-          x_hi,
-          y_lo,
-          y_hi,
-          blk_h >> 1,
-          blk_w >> 1,
-          &fs.input_hres,
-          &rec.input_hres,
-          &mut best_mv,
-          &mut lowest_cost,
-          &po,
-          1,
-          fi.sequence.bit_depth,
-          lambda,
-          [MotionVector::default(); 2],
-          fi.allow_high_precision_mv
-        );
+          for predictor in &mut predictors {
+            predictor.row = predictor.row >> 1;
+            predictor.col = predictor.col >> 1;
+          }
+
+          diamond_me_search(
+            fi, &po,
+            &fs.input_hres, &rec.input_hres,
+            &predictors, fi.sequence.bit_depth,
+            global_mv, lambda,
+            mvx_min >> 1, mvx_max >> 1, mvy_min >> 1, mvy_max >> 1,
+            blk_w >> 1, blk_h >> 1,
+            &mut best_mv, &mut lowest_cost,
+            &mut None, 0
+          );
+        } else {
```

Comment:
Make them two functions and push the branch as up as possible in the call tree.

Reply:
I extracted the if clause from the loop in the last version of this PR.
I have no issue making two functions. I however do not see in practice how to do it properly.